identity/src/main/java/com/android/identity/credential/Credential.java

```diff
@@ -0,0 +1,1074 @@
+        saveCredential();
+    }
+
+    /**
+     * Gets application specific data.
+     *
+     * <p>Gets data previously stored with {@link #getApplicationData(String)}.
+     *
+     * @param key the key for the data.
+     * @return the value or {@code null} if there is no value for the given key.
+     */
+    public @Nullable byte[] getApplicationData(@NonNull String key) {
+        return mApplicationData.get(key);
+    }
+
+    /**
+     * Gets application specific data as a string.
+     *
+     * <p>Like {@link #getApplicationData(String)} but decodes the value as
+     * an UTF-8 string before returning it.
+     *
+     * @param key the key for the data.
+     * @return the value or {@code null} if there is no value for the given key.
+     */
+    public @Nullable String getApplicationDataString(@NonNull String key) {
+        byte[] value = mApplicationData.get(key);
+        if (value == null) {
+            return null;
+        }
+        return new String(value, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Finds a suitable authentication key to use.
+     *
+     * @param now Pass current time to ensure that the selected slot's validity period or
+     *   {@code null} to not consider validity times.
+     * @return An authentication key which can be used for signing or {@code null} if none was found.
+     */
+    public @Nullable AuthenticationKey findAuthenticationKey(@Nullable Timestamp now) {
+
+        AuthenticationKey candidate = null;
+        for (AuthenticationKey authenticationKey : mAuthenticationKeys) {
+            // If current time is passed...
+            if (now != null) {
+                // ... ignore slots that aren't yet valid
+                if (now.toEpochMilli() < authenticationKey.getValidFrom().toEpochMilli()) {
+                    continue;
+                }
+                // .. ignore slots that aren't valid anymore
+                if (now.toEpochMilli() > authenticationKey.getValidUntil().toEpochMilli()) {
+                    continue;
+                }
+            }
+            // If we already have a candidate, prefer this one if its usage count is lower
+            if (candidate != null) {
+                if (authenticationKey.getUsageCount() < candidate.getUsageCount()) {
+                    candidate = authenticationKey;
+                }
+            } else {
+                candidate = authenticationKey;
+            }
+        }
+        return candidate;
+    }
+
+    /**
+     * A certified authentication key.
+     *
+     * <p>To create an instance of this type, an application must first use
+     * {@link Credential#createPendingAuthenticationKey(KeystoreEngine.CreateKeySettings, AuthenticationKey)}
+     * to create a {@link PendingAuthenticationKey} and after issuer certification
+     * has been received it can be upgraded to a {@link AuthenticationKey}.
+     */
+    public static class AuthenticationKey {
+        String mAlias;
+        int mUsageCount;
```

Comment:
I feel like this should be `private` so it can't be arbitrarily modified outside `increaseUsageCount`

Reply:
Yeah, a bunch of these can be private I think. Doing that.